private/pkg/protoencoding/resolver.go

```diff
@@ -40,50 +40,58 @@ func newResolver(fileDescriptors ...protodescriptor.FileDescriptor) (Resolver, e
 	types := &protoregistry.Types{}
 	var rangeErr error
 	files.RangeFiles(func(fileDescriptor protoreflect.FileDescriptor) bool {
-		if err := addMessagesToTypes(types, fileDescriptor.Messages()); err != nil {
+		if err := registerDescriptors(types, fileDescriptor); err != nil {
 			rangeErr = err
 			return false
 		}
-		if err := addExtensionsToTypes(types, fileDescriptor.Extensions()); err != nil {
-			rangeErr = err
-			return false
-		}
-		// There is no way to do register enum, and it is not used
-		// https://github.com/golang/protobuf/issues/1065
-		// https://godoc.org/google.golang.org/protobuf/types/dynamicpb does not have NewEnumType
 		return true
 	})
 	if rangeErr != nil {
 		return nil, rangeErr
 	}
-	return types, nil
+	return &resolver{Files: files, Types: types}, nil
+}
+
+type resolver struct {
+	*protoregistry.Files
+	*protoregistry.Types
+}
+
+type descriptorContainer interface {
+	Messages() protoreflect.MessageDescriptors
+	Enums() protoreflect.EnumDescriptors
+	Extensions() protoreflect.ExtensionDescriptors
 }
 
-func addMessagesToTypes(types *protoregistry.Types, messageDescriptors protoreflect.MessageDescriptors) error {
-	messagesLen := messageDescriptors.Len()
-	for i := 0; i < messagesLen; i++ {
+func registerDescriptors(types *protoregistry.Types, container descriptorContainer) error {
+	messageDescriptors := container.Messages()
+	for i, messagesLen := 0, messageDescriptors.Len(); i < messagesLen; i++ {
 		messageDescriptor := messageDescriptors.Get(i)
 		if err := types.RegisterMessage(dynamicpb.NewMessageType(messageDescriptor)); err != nil {
 			return err
 		}
-		if err := addMessagesToTypes(types, messageDescriptor.Messages()); err != nil {
+		// nested types, too
+		if err := registerDescriptors(types, messageDescriptor); err != nil {
 			return err
 		}
-		if err := addExtensionsToTypes(types, messageDescriptor.Extensions()); err != nil {
+	}
+
+	enumDescriptors := container.Enums()
+	for i, enumsLen := 0, enumDescriptors.Len(); i < enumsLen; i++ {
+		enumDescriptor := enumDescriptors.Get(i)
+		if err := types.RegisterEnum(dynamicpb.NewEnumType(enumDescriptor)); err != nil {
```

Comment:
Nice, the issues linked before seem closed as a wontfix but seems like this made it after all: https://go-review.googlesource.com/c/protobuf/+/237219